### PR TITLE
DOMA-2807 ticket hint

### PR DESCRIPTION
**Pull request by nomerdvadcatpyat**
Ticket hint - information related to a property or group of properties in organization displayed in the ticket form and in the ticket card

[Figma layouts](https://www.figma.com/file/P827yyueCLRT6o4gyNtzY6/CRM-(v1.doma.ai)?node-id=18954%3A106335)

### CODE REVIEW

File: apps/condo/domains/ticket/components/TicketHint/BaseTicketHintForm.tsx [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+import Checkbox from '@condo/domains/common/components/antd/Checkbox'
+import Select from '@condo/domains/common/components/antd/Select'
+import { FormWithAction } from '@condo/domains/common/components/containers/FormList'
+import { GraphQlSearchInput } from '@condo/domains/common/components/GraphQlSearchInput'
+import { Loader } from '@condo/domains/common/components/Loader'
+import { colors } from '@condo/domains/common/constants/style'
+import { useValidations } from '@condo/domains/common/hooks/useValidations'
+
+import { searchOrganizationProperty } from '../../utils/clientSchema/search'
+import { Property } from '../../../property/utils/clientSchema'
```
Comment (sitozzz) [inline apps/condo/domains/ticket/components/TicketHint/BaseTicketHintForm.tsx:20]: Plz use absolute import

File: apps/condo/domains/ticket/components/TicketHint/SettingsContent.tsx [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+const SORTABLE_PROPERTIES = ['name']
+const TICKET_HINTS_DEFAULT_SORT_BY = ['createdAt_DESC']
+
+const StyledTable = styled(Table)`
+  .ant-table-cell-ellipsis {
+    white-space: inherit;
+  }
+`
+
+const GUTTER_0_40: [Gutter, Gutter] = [0, 40]
```
Comment (sitozzz) [inline apps/condo/domains/ticket/components/TicketHint/SettingsContent.tsx:37]: I don't think the name of a style constant should contain its value. If you need to change it in the future, then you will have to change the name of the constant

File: apps/condo/domains/common/components/settings/SettingsPageContent.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
+import React, { CSSProperties, useCallback, useMemo } from 'react'
+import { Tabs } from 'antd'
+import { useRouter } from 'next/router'
+
+import { parseQuery } from '../../utils/tables.utils'
```
Comment (sitozzz) [inline apps/condo/domains/common/components/settings/SettingsPageContent.tsx:5]: absolute import

File: apps/condo/domains/ticket/components/TicketHint/CreateTicketHintForm.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import { useRouter } from 'next/router'
+import React, { useMemo } from 'react'
+
+import { useOrganization } from '@core/next/organization'
+import { useIntl } from '@core/next/intl'
+
+import ActionBar from '@condo/domains/common/components/ActionBar'
+import { Button } from '@condo/domains/common/components/Button'
+
+import { TicketHint } from '../../utils/clientSchema'
```
Comment (sitozzz) [inline apps/condo/domains/ticket/components/TicketHint/CreateTicketHintForm.tsx:11]: absolute import

File: apps/condo/domains/ticket/components/TicketHint/TicketIdHint.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import { InfoCircleOutlined } from '@ant-design/icons'
+import styled from '@emotion/styled'
+import { Alert, Typography } from 'antd'
+import Link from 'next/link'
+
+import { colors, fontSizes } from '@condo/domains/common/constants/style'
+import { useIntl } from '@core/next/intl'
+import { CSSProperties } from 'react'
+
+import { TicketHint } from '../../utils/clientSchema'
```
Comment (sitozzz) [inline apps/condo/domains/ticket/components/TicketHint/TicketIdHint.tsx:10]: absolute import

File: apps/condo/domains/ticket/schema/TicketHint.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        name: {
+            schemaDoc: 'Hint name',
+            type: Text,
+        },
+
+        properties: {
+            schemaDoc: 'Properties to which this help applies',
+            type: Relationship,
+            ref: 'Property',
+            many: true,
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketHint.js:28]: Do not use `many:true`. We need to re-design it

File: apps/condo/domains/ticket/schema/TicketHint.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        },
+
+        content: {
+            schemaDoc: 'Textual content of help in HTML format',
+            type: Text,
+            isRequired: true,
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketHint.js:34]: validation? normalization?

File: apps/condo/pages/property/[id]/hint.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    const propertyId = get(router, ['query', 'id'], null)
+
+    const { loading: propertyLoading, obj: property, error } = useObject({ where: { id: propertyId } })
+    const { loading: ticketHintLoading, obj: ticketHint } = TicketHint.useObject({
+        where: {
+            properties_some: { id: propertyId },
+        },
+    })
+
+    const htmlContent = useMemo(() => ({
```
Comment (pahaz) [inline apps/condo/pages/property/[id]/hint.tsx:36]: what about `TicketHintContent` ?

File: apps/condo/domains/ticket/schema/TicketHint.js (hunk 7 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/ticket/schema/TicketHint.js:27]: added normalization using the xss library, wrote a test

**Comment by AntonAL** [comment]
Information related to a property or group of properties seems to be not related with Tickets. You tied the name TicketHint to it's display place, but not to a logical relations. Are you sure that this piece of information will always be displayed only in ticket page?

File: apps/condo/domains/ticket/components/TicketHint/SettingsContent.tsx [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+import { useOrganization } from '@core/next/organization'
+import { SortTicketHintsBy } from '@app/condo/schema'
+import ActionBar from '@condo/domains/common/components/ActionBar'
+import { Button } from '@condo/domains/common/components/Button'
+import { useQueryMappers } from '@condo/domains/common/hooks/useQueryMappers'
+
+import { useTicketHintTableColumns } from '@condo/domains/ticket/hooks/useTicketHintTableColumns'
+import { useTicketHintTableFilters } from '@condo/domains/ticket/hooks/useTicketHintTableFilters'
+import { TicketHint } from '../../utils/clientSchema'
+import { IFilters } from '../../utils/helpers'
```
Comment (sitozzz) [inline apps/condo/domains/ticket/components/TicketHint/SettingsContent.tsx:26]: Absolute import :)

File: apps/condo/domains/ticket/components/TicketHint/TicketIdHintCard.tsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+import styled from '@emotion/styled'
+import { Alert, Typography } from 'antd'
+import { get } from 'lodash'
+import Link from 'next/link'
+
+import { colors, fontSizes } from '@condo/domains/common/constants/style'
+import { useIntl } from '@core/next/intl'
+import { CSSProperties, useMemo } from 'react'
+
+import { TicketHint, TicketHintProperty } from '../../utils/clientSchema'
```
Comment (sitozzz) [inline apps/condo/domains/ticket/components/TicketHint/TicketIdHintCard.tsx:11]: Absolute import

File: apps/condo/domains/ticket/utils/clientSchema/TicketHintProperty.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+ * Generated by `createschema ticket.TicketHintProperty 'organization:Relationship:Organization:CASCADE;ticketHint:Relationship:TicketHint:CASCADE; property:Relationship:Property:SET_NULL;'`
+ */
+
+import { pick, get } from 'lodash'
+
+import { getClientSideSenderInfo } from '@condo/domains/common/utils/userid.utils'
+import { generateReactHooks } from '@condo/domains/common/utils/codegeneration/generate.hooks'
+
+import { TicketHintProperty as TicketHintPropertyGQL } from '@condo/domains/ticket/gql'
+import { TicketHintProperty, TicketHintPropertyUpdateInput, QueryAllTicketHintPropertiesArgs } from '../../../../schema'
```
Comment (sitozzz) [inline apps/condo/domains/ticket/utils/clientSchema/TicketHintProperty.ts:11]:
```suggestion
import { TicketHintProperty, TicketHintPropertyUpdateInput, QueryAllTicketHintPropertiesArgs } from '@app/condo/schema'
```

File: apps/condo/pages/settings/hint/[id]/index.tsx [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+    const router = useRouter()
+    const { link } = useOrganization()
+    const canManageTicketHints = useMemo(() => get(link, ['role', 'canManageTicketHints']), [link])
+
+    const hintId = get(router, ['query', 'id'], null)
+    const { loading, obj: ticketHint } = TicketHint.useObject({
+        where: { id: hintId },
+    }, {
+        fetchPolicy: 'network-only',
```
Comment (sitozzz) [inline apps/condo/pages/settings/hint/[id]/index.tsx:51]: Why not to use cache here?

File: apps/condo/domains/ticket/components/TicketHint/TicketIdHintCard.tsx [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  & svg, & > .ant-alert-content > .ant-alert-message {
+   color: ${colors.black}; 
+  }
+`
+
+const TEXT_STYLES: CSSProperties = { color: colors.black }
+const LINK_STYLES: CSSProperties = { ...TEXT_STYLES, position: 'relative', bottom: '-7px', fontSize: fontSizes.content }
+const TICKET_HINT_CONTENT_STYLES: CSSProperties = { maxHeight: '5em' }
+
+export const TicketIdHintCard = ({ propertyId }) => {
```
Comment (AntonAL) [inline apps/condo/domains/ticket/components/TicketHint/TicketIdHintCard.tsx:27]: Name overlaps with `Card` component from Ant.
`TicketId…` confuses in terms of displaying id of a ticket.

File: apps/condo/domains/ticket/schema/TicketHint.js [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+const { Text, Relationship } = require('@keystonejs/fields')
+const { GQLListSchema } = require('@core/keystone/schema')
+const { historical, versioned, uuided, tracked, softDeleted } = require('@core/keystone/plugins')
+const { dvAndSender } = require('@condo/domains/common/schema/plugins/dvAndSender')
+const access = require('@condo/domains/ticket/access/TicketHint')
+const { ORGANIZATION_OWNED_FIELD } = require('@condo/domains/organization/schema/fields')
+const xss = require('xss')
+
+const TicketHint = new GQLListSchema('TicketHint', {
+    schemaDoc: 'Help information related to a property or group of properties',
```
Comment (AntonAL) [inline apps/condo/domains/ticket/schema/TicketHint.js:14]: Could you bring some examples of what can be presented in information text?

```suggestion
    schemaDoc: 'Textual information in free format related to a property or group of properties, for example…',
```

File: apps/condo/domains/ticket/schema/TicketHintProperty.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+const { Relationship } = require('@keystonejs/fields')
+
+const { GQLListSchema } = require('@core/keystone/schema')
+const { historical, versioned, uuided, tracked, softDeleted } = require('@core/keystone/plugins')
+
+const { dvAndSender } = require('@condo/domains/common/schema/plugins/dvAndSender')
+const access = require('@condo/domains/ticket/access/TicketHintProperty')
+const { ORGANIZATION_OWNED_FIELD } = require('@condo/domains/organization/schema/fields')
+
+const TicketHintProperty = new GQLListSchema('TicketHintProperty', {
```
Comment (AntonAL) [inline apps/condo/domains/ticket/schema/TicketHintProperty.js:14]: Is it not a redundant schema in your case? AFAIK, Keystone can generate a schema `Ticket_properties_many` if you specify `many: true` for a `TicketHint`.

File: apps/condo/domains/ticket/components/TicketHint/FormTicketHintCard.tsx [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+import { colors } from '@condo/domains/common/constants/style'
+import { TicketHint, TicketHintProperty } from '@condo/domains/ticket/utils/clientSchema'
+
+import { TicketHintContent } from './TicketHintContent'
+import { get } from 'lodash'
+
+const ROW_STYLES: CSSProperties = { overflow: 'hidden', backgroundColor: colors.backgroundLightGrey, padding: '20px', borderRadius: '12px' }
+const LINK_STYLES: CSSProperties = { color: 'black', position: 'relative', bottom: '-8px' }
+
+const SMALL_VERTICAL_GUTTER: [Gutter, Gutter] = [0, 10]
```
Comment (AntonAL) [inline apps/condo/domains/ticket/components/TicketHint/FormTicketHintCard.tsx:17]: We have a `MAIN_ROW_GUTTER` constant with the same value. Can it be reused in this case?

File: apps/condo/domains/ticket/utils/clientSchema/Renders.tsx [hunk 17]
```diff
@@ -186,4 +188,27 @@ export const getTicketClientNameRender = (search: FilterValue) => {
 
         return getTableCellRenderer(search, false, null, null, null, clientName)(trimmedClientName)
     }
+}
+
+const HINT_STYLES: CSSProperties = { maxHeight: '6.5em', maxWidth: '300px', overflow: 'hidden', wordBreak: 'break-word', whiteSpace: 'inherit' }
+
+export const getTicketHintRender = (search: FilterValue) => {
```
Comment (AntonAL) [inline apps/condo/domains/ticket/utils/clientSchema/Renders.tsx:195]: According to implementation of this function, it renders html content.
Maybe to create a React component `HtmlContent` in `common` domain?

File: apps/condo/pages/settings/index.tsx [hunk 18]
```diff
@@ -1,44 +1,50 @@
...
-    const RolesAndAccessesTitle = intl.formatMessage({ id: 'RolesAndAccess' })
+    const HintTitle = intl.formatMessage({ id: 'Hint' })
     const SubscriptionTitle = intl.formatMessage({ id: 'Subscription' })
 
     const hasSubscriptionFeature = hasFeature('subscription')
-
-    const router = useRouter()
-    const { tab } = parseQuery(router.query)
-
     const availableTabs = useMemo(() => {
```
Comment (AntonAL) [inline apps/condo/pages/settings/index.tsx:27]:
```suggestion
    const tabKeysToDisplay = useMemo(() => {
```

File: apps/condo/pages/settings/index.tsx [hunk 19]
```diff
@@ -1,44 +1,50 @@
...
         const result = ALWAYS_AVAILABLE_TABS
         if (hasSubscriptionFeature) result.push('subscription')
         return result
     }, [hasSubscriptionFeature])
 
-    const defaultTab = availableTabs.includes(tab) ? tab : undefined
+    const SETTINGS_TABS: SettingsTabType[] = useMemo(() => [
```
Comment (AntonAL) [inline apps/condo/pages/settings/index.tsx:33]:
```suggestion
    const settingsTabs: SettingsTabType[] = useMemo(() => [
```

File: apps/condo/lang/en/en.json [hunk 20]
```diff
@@ -611,6 +613,17 @@
   "pages.condo.meter.EndDate": "End date",
   "pages.condo.meter.EnterMeterNumber": "Enter number",
   "pages.condo.meter.EnterPlace": "Enter place",
+  "pages.condo.settings.hint.propertyHint": "Property hint",
+  "pages.condo.settings.hint.addAllProperties": "Add all properties",
+  "pages.condo.settings.hint.createTicketHint": "Create ticket hint",
+  "pages.condo.settings.hint.newTicketHint": "New hint",
+  "pages.condo.settings.hint.editTicketHint": "Edit hint",
+  "pages.condo.settings.hint.alert.title": "For some properties of this organization there are already hints",
```
Comment (AntonAL) [inline apps/condo/lang/en/en.json:621]:
```suggestion
  "pages.condo.settings.hint.alert.title": "Some properties of this organization already have hints",
```

Comment (AntonAL) [inline apps/condo/lang/en/en.json:621]: If it is a validation, maybe to rename keys accordingly?

```suggestion
  "pages.condo.settings.hint.validation.alreadyPresented": "For some properties of this organization there are already hints",
```

File: apps/condo/domains/ticket/schema/TicketHint.js (hunk 6 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/ticket/schema/TicketHint.js]: done

File: apps/condo/pages/settings/hint/[id]/index.tsx (hunk 12 above)
Comment (nomerdvadcatpyat) [reply apps/condo/pages/settings/hint/[id]/index.tsx:52]: removed it

File: apps/condo/domains/common/components/settings/SettingsPageContent.tsx [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+import { SettingsTab, SettingsTabs } from './Tabs'
+
+export type SettingsTabType = {
+    key: string,
+    title: string,
+    content?: React.ReactElement,
+}
```
[the remark this answers is not in the document]
Comment (nomerdvadcatpyat) [reply apps/condo/domains/common/components/settings/SettingsPageContent.tsx:13]: moved in Tabs module

File: apps/condo/domains/ticket/schema/TicketHintProperty.js (hunk 15 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/ticket/schema/TicketHintProperty.js]: No, it's better to create an explicit model than to use `many: true`. `many: true` results in very expensive queries, we want to move away from this in the future.

File: apps/condo/domains/ticket/utils/clientSchema/Renders.tsx (hunk 17 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/ticket/utils/clientSchema/Renders.tsx]: Added `HtmlContent` in common domain and reuse it here

File: apps/condo/lang/en/en.json (hunk 20 above)
Comment (nomerdvadcatpyat) [reply apps/condo/lang/en/en.json]: No, it's not validation, it's helper alert

File: apps/condo/domains/ticket/access/TicketPropertyHint.js [hunk 22]
```diff
@@ -0,0 +1,56 @@
+// @ts-nocheck
```
Comment (sitozzz) [inline apps/condo/domains/ticket/access/TicketPropertyHint.js:1]: May this marker be removed?

**Review by sitozzz — APPROVED** [review]

File: apps/condo/domains/ticket/schema/TicketPropertyHintProperty.js [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+
+const { GQLListSchema } = require('@core/keystone/schema')
+const { historical, versioned, uuided, tracked, softDeleted } = require('@core/keystone/plugins')
+
+const { dvAndSender } = require('@condo/domains/common/schema/plugins/dvAndSender')
+const access = require('@condo/domains/ticket/access/TicketPropertyHintProperty')
+const { ORGANIZATION_OWNED_FIELD } = require('@condo/domains/organization/schema/fields')
+
+const TicketPropertyHintProperty = new GQLListSchema('TicketPropertyHintProperty', {
+    schemaDoc: 'Sets to which property the ticketPropertyHint belongs',
```
Comment (AntonAL) [inline apps/condo/domains/ticket/schema/TicketPropertyHintProperty.js:15]:
```suggestion
    schemaDoc: 'Join entity for many-to-many relations, that determines a Property to which the TicketPropertyHint belongs',
```

File: apps/condo/domains/ticket/utils/clientSchema/search.ts [hunk 24]
```diff
@@ -101,6 +101,15 @@ const GET_ALL_CONTACTS_QUERY = gql`
     }
 `
 
+const GET_ALL_PROPERTY_HINTS_TO_PROPERTIES_QUERY = gql`
+    query selectPropertyHintToProperty ($where: TicketHintPropertyWhereInput, $orderBy: String, $first: Int, $skip: Int) {
```
Comment (AntonAL) [inline apps/condo/domains/ticket/utils/clientSchema/search.ts:105]:
```suggestion
    query selectTicketPropertyHintProperty ($where: TicketHintPropertyWhereInput, $orderBy: String, $first: Int, $skip: Int) {
```

File: apps/condo/domains/ticket/utils/clientSchema/search.ts [hunk 25]
```diff
@@ -101,6 +101,15 @@ const GET_ALL_CONTACTS_QUERY = gql`
     }
 `
 
+const GET_ALL_PROPERTY_HINTS_TO_PROPERTIES_QUERY = gql`
```
Comment (AntonAL) [inline apps/condo/domains/ticket/utils/clientSchema/search.ts:104]:
```suggestion
const GET_ALL_PROPERTY_HINT_PROPERTIES_QUERY = gql`
```

File: apps/condo/domains/ticket/utils/clientSchema/search.ts [hunk 26]
```diff
@@ -101,6 +101,15 @@ const GET_ALL_CONTACTS_QUERY = gql`
...
+    query selectPropertyHintToProperty ($where: TicketHintPropertyWhereInput, $orderBy: String, $first: Int, $skip: Int) {
+        objs: allTicketHintProperties(where: $where, orderBy: $orderBy, first: $first, skip: $skip) {
```
Comment (AntonAL) [inline apps/condo/domains/ticket/utils/clientSchema/search.ts:106]: Is it a correct naming?

File: apps/condo/domains/ticket/utils/clientSchema/search.ts (hunk 24 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/ticket/utils/clientSchema/search.ts:105]: it's redundant const, removed it

File: apps/condo/domains/ticket/utils/clientSchema/search.ts (hunk 26 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/ticket/utils/clientSchema/search.ts:106]: it's redundant const, removed it

File: .eslintignore [hunk 27]
```diff
@@ -12,3 +12,4 @@ apps/sberhack
 # Ignore other files
 package.json
 README.md
+**/public/tinymce
```
Comment (pahaz) [inline .eslintignore:15]: are you sure that we need to ignore it?

File: apps/condo/domains/common/utils/xss.js [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
+    whiteList,
+    css: {
+        whiteList: {
+            'text-align': true,
+            'background-color': true,
+        },
+    },
+})
+
+module.exports = { xss }
```
Comment (pahaz) [inline apps/condo/domains/common/utils/xss.js:19]: it's bettert to just export something like `sanitizeXss(html)`. It's help us to change the XSS chack implementation in a future

File: apps/condo/domains/property/schema/Property.js [hunk 29]
```diff
@@ -347,6 +344,23 @@ const Property = new GQLListSchema('Property', {
...
+                        deletedAt: null,
+                    })
+
+                    for (const ticketPropertyHintProperty of ticketPropertyHintProperties) {
+                        await TicketPropertyHintProperty.update(context, ticketPropertyHintProperty.id, {
+                            deletedAt: now,
+                            dv, sender,
+                        })
+                    }
+                }
```
Comment (pahaz) [inline apps/condo/domains/property/schema/Property.js:363]: it's better to extract this logit to some function to increase the readability

File: apps/condo/domains/common/utils/xss.js (hunk 28 above)
Comment (pahaz) [reply apps/condo/domains/common/utils/xss.js:19]: And write some tests for this util.

File: apps/condo/pages/property/[id]/hint.tsx [hunk 30]
```diff
@@ -0,0 +1,84 @@
+/** @jsx jsx */
+import { Gutter } from 'antd/es/grid/row'
+import { get } from 'lodash'
+import React, { CSSProperties, useMemo } from 'react'
+import { useIntl } from '@core/next/intl'
+import { useRouter } from 'next/router'
+import { jsx } from '@emotion/core'
+import Head from 'next/head'
+import xss from 'xss'
```
Comment (pahaz) [inline apps/condo/pages/property/[id]/hint.tsx:9]: no usage?

File: apps/condo/domains/ticket/components/TicketPropertyHint/BaseTicketPropertyHintForm.tsx [hunk 31]
```diff
@@ -0,0 +1,282 @@
...
+}
+const HINT_CONTENT_FIELD_LAYOUT_PROPS = {
+    wrapperCol: { span: 0 },
+}
+
+const {
+    publicRuntimeConfig,
+} = getConfig()
+
+const { TinyMceApiKey } = publicRuntimeConfig
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/TicketPropertyHint/BaseTicketPropertyHintForm.tsx:84]: move it to runtime

**Review by pahaz — APPROVED** [review]

File: .eslintignore (hunk 27 above)
Comment (nomerdvadcatpyat) [reply .eslintignore:15]: yes

File: apps/condo/pages/property/[id]/hint.tsx (hunk 30 above)
Comment (nomerdvadcatpyat) [reply apps/condo/pages/property/[id]/hint.tsx:9]: removed

**Pull request merged**